src/cargo/core/compiler/standard_lib.rs

```diff
@@ -31,13 +34,43 @@ pub fn parse_unstable_flag(value: Option<&str>) -> Vec<String> {
     crates.into_iter().map(|s| s.to_string()).collect()
 }
 
+pub(crate) fn std_crates(config: &Config, units: Option<&[Unit]>) -> Option<Vec<String>> {
+    let crates = config.cli_unstable().build_std.as_ref()?.clone();
+
+    // Only build libtest if it looks like it is needed.
+    let mut crates = crates.clone();
+    // If we know what units we're building, we can filter for libtest depending on the jobs.
+    if let Some(units) = units {
+        if units
+            .iter()
+            .any(|unit| unit.mode.is_rustc_test() && unit.target.harness())
+        {
+            // Only build libtest when libstd is built (libtest depends on libstd)
+            if crates.iter().any(|c| c == "std") && !crates.iter().any(|c| c == "test") {
+                crates.push("test".to_string());
+            }
+        }
+    } else {
+        // We don't know what jobs are going to be run, so download libtest just in case.
+        crates.push("test".to_string())
+    }
+
+    Some(crates)
+}
+
 /// Resolve the standard library dependencies.
 pub fn resolve_std<'cfg>(
     ws: &Workspace<'cfg>,
     target_data: &RustcTargetData<'cfg>,
-    requested_targets: &[CompileKind],
+    build_config: &BuildConfig,
     crates: &[String],
 ) -> CargoResult<(PackageSet<'cfg>, Resolve, ResolvedFeatures)> {
+    if build_config.build_plan {
+        ws.config()
+            .shell()
+            .warn("-Zbuild-std does not currently fully support --build-plan")?;
+    }
+
```

Comment:
:shrug: in case Cargo ever supports it for things like `cargo check` or cargo doc. I don't think it does currently, but it doesn't do any harm to support this in case it does.